[multistage] hintable option framework & dynamic broadcast hint

walterddr opened the pull request:
Details
===
- add hintable option framework
  - indicates the hint applicable node (`aggOptions`, `joinOptions` are added for now)
  - each hint support arbitrary k-v options (e.g. `joinOptions(join_strategy='dynamic_broadcast', colocated_by_join_keys='true')`
- also added the dynamic_broadcast hint and rel optimizer rule 
  - this helps execute queries similar to ([IN_SUBQUERY](https://docs.pinot.apache.org/users/user-guide-query/filtering-with-idset#in_subquery))
- also sanitized agg hint options and tests.
  - also demonstrated that how mixed hint options work (join options and agg options)


Examples
===
- hint with hint name:
```
SELECT /*+ skipLeafStageGroupByAggregation */ 
  a, sum(b) 
FROM tbl 
GROUP BY a
```
- hint with hint options on the specific node
```
SELECT /*+ joinOptions(join_strategy='dynamic_broadcast', colocated_by_join_keys='true') */
  a, b
FROM tbl
WHERE b BETWEEN 10 AND 100
  AND c IN (SELECT d FROM tbl2 WHERE e > 10)
```

Extensibility
===
with this hint strategy framework we can extend easily.

1. adding more hintable node, such as hint on WindowNode
    - we can simply add a `windowOptions` type to `PinotHintOptions` class
2. adding more generic option keyword, such as support `partition_by='col1,col2'` on aggregate and table node. 
    - we can simply add generic k-v option keys to the `PinotHintOptions.aggOptions` sub-class to indicate that a group by should be partitioned by "col1,col2" instead of the group set.
3. adding some completely unknown keyword 
    - the entire options framework is a `Map<String, String>` basis, so we can directly extract the string value, such as on a plugged-in ruleset from some fork. we can simply extract it via String key to get the String value

walterddr commented:
CC 
- @vvivekiyer and @somandal who visited the hintable framework previously on aggregate and window functions
- @ankitsultana for the colocated options.

xiangfu0 reviewed (approved):
lgtm

walterddr commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/hint/PinotHintOptions.java:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  public static class AggregateOptions {
+    public static final String IS_PARTITIONED_BY_GROUP_BY_KEYS = "partitioned_by_group_by_keys";
```
That's a different option i tentatively name it as: `partition_keys='col1,col2,col3` and it doesn't have to apply to group-by but can also apply to table scan node

walterddr commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:
```diff
@@ -88,8 +92,11 @@ public boolean matches(RelOptRuleCall call) {
+      return !PinotHintStrategyTable.containsHint(hints, PinotHintStrategyTable.INTERNAL_AGG_INTERMEDIATE_STAGE)
+          && !PinotHintStrategyTable.containsHint(hints, PinotHintStrategyTable.INTERNAL_AGG_FINAL_STAGE)
+          && !PinotHintStrategyTable.containsHintOption(hints, PinotHintOptions.AGGREGATE_HINT_OPTIONS,
+          PinotHintOptions.AggregateOptions.IS_PARTITIONED_BY_GROUP_BY_KEYS);
```
it will not be considered. but good point will add a test

kishoreg commented:
syntax for hints seems a bit verbose.. can we follow spark or postgres model

https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-hints.html

walterddr commented:
> syntax for hints seems a bit verbose.. can we follow spark or postgres model
> 
> https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-hints.html

There will always be the option to make top-level hint instead of going through the option k-v route. it will definitely be less verbose
However the current arbitrary key-value options support will make it more flexible to extend in both plan time and runtime optimization code as it doesn't require a fixed set of hints registered

ankitsultana commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/hint/PinotHintOptions.java:
```diff
@@ -0,0 +1,44 @@
+
+  public static class JoinHintOptions {
+    public static final String JOIN_STRATEGY = "join_strategy";
+    public static final String IS_COLOCATED_BY_JOIN_KEYS = "colocated_by_join_keys";
```
Are we planning to keep these hints long-term or are these only until we implement auto-detection for colocation?

walterddr replied:
technically speaking we can keep them --> e.g. even the data are not colocated it doesn't stop the user from "just join those that are colocated on the same server and see the result", it is definitely useful to test the data colocation.

ankitsultana reviewed (approved):
Looks good from our side.

siddharthteotia commented:
I am also reviewing this today

walterddr commented:
addressed all comments and we have a basic understanding on the hint option framework.

will merge and will address any follow up comments later. thx
